program-runtime/src/guest_transaction.rs

```diff
@@ -0,0 +1,301 @@
+
+        let scratchpad = transaction_context.get_return_data();
+
+        guest_transaction.return_data_scratchpad = ReturnDataScratchpad {
+            pubkey: *scratchpad.0,
+            slice: GuestSliceReference {
+                pointer: RETURN_DATA_SCRATCHPAD_ADDRESS,
+                length: scratchpad.1.len() as u64,
+            },
+        };
+
+        guest_transaction.cpi_scratchpad = GuestSliceReference {
+            pointer: CPI_SCRATCHPAD_ADDRESS,
+            length: 0,
+        };
+
+        guest_transaction.instruction_idx = 0;
+        guest_transaction.instruction_num = num_instructions as u64;
+        guest_transaction.accounts_no = transaction_context.get_number_of_accounts() as u64;
+
+        // SAFETY: The memory region is large enough to contain a GuestTransactionContext and an
+        // array of `GuestTransactionAccount`
+        let transaction_accounts = unsafe {
+            let ptr = memory
+                .as_mut_ptr()
+                .add(size_of::<GuestTransactionContext>());
+            slice::from_raw_parts_mut(
+                ptr as *mut GuestTransactionAccount,
+                guest_transaction.accounts_no as usize,
+            )
+        };
+
+        for index_in_transaction in 0..transaction_context.get_number_of_accounts() {
+            let transaction_account = transaction_context
+                .accounts()
+                .try_borrow(index_in_transaction)
+                .unwrap();
+
+            let account_ref = transaction_accounts
+                .get_mut(index_in_transaction as usize)
+                .unwrap();
+            account_ref.pubkey = *transaction_context
+                .get_key_of_account_at_index(index_in_transaction)
+                .unwrap();
+            account_ref.owner = *transaction_account.owner();
+            account_ref.lamports = transaction_account.lamports();
+            let vm_data_addr = ACCOUNT_DATA_VM_ADDRESS
+                .saturating_add(MM_REGION_SIZE.saturating_mul(index_in_transaction as u64));
+            account_ref.data = GuestSliceReference {
+                pointer: vm_data_addr,
+                length: transaction_account.data().len() as u64,
+            };
+        }
+
+        // SAFETY: The vector has been allocated with at least `size` bytes.
+        unsafe {
+            memory_vec.set_len(size);
+        }
+
+        RuntimeGuestTransaction {
+            data: memory_vec.into_boxed_slice(),
+        }
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        &self.data
+    }
+
+    pub fn set_instruction_index(&mut self, index: usize) {
+        // SAFETY: We assume the transaction was created using `RuntimeGuestTransaction::new`, which
+        // guarantees the safety of size constraints and contents.
+        let context = unsafe { &mut *(self.data.as_mut_ptr() as *mut GuestTransactionContext) };
+
+        context.instruction_idx = index as u64;
+    }
+}
```

Comment:
I plan on adding functions to retrieve `GuestTransactionContext` and `&[GuestTransactionAccount]` directly. Since they were not necessary for this PR, I left them out.